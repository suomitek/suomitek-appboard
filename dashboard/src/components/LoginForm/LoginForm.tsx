import { Location } from "history";
import * as React from "react";
import { Lock } from "react-feather";
import { Redirect } from "react-router";

import LoadingWrapper from "../../components/LoadingWrapper";
import "./LoginForm.css";

export interface ILoginFormProps {
  cluster: string;
  authenticated: boolean;
  authenticating: boolean;
  authenticationError: string | undefined;
  oauthLoginURI: string;
  authenticate: (cluster: string, token: string) => any;
  checkCookieAuthentication: (cluster: string) => void;
  location: Location;
}

interface ILoginFormState {
  token: string;
}

class LoginForm extends React.Component<ILoginFormProps, ILoginFormState> {
  public state: ILoginFormState = { token: "" };

  public componentDidMount() {
    if (this.props.oauthLoginURI) {
      this.props.checkCookieAuthentication(this.props.cluster);
    }
  }

  public render() {
    if (this.props.authenticating) {
      return <LoadingWrapper />;
    }
    if (this.props.authenticated) {
      const { from } = (this.props.location.state as any) || { from: { pathname: "/" } };
      return <Redirect to={from} />;
    }

    return (
      <section className="LoginForm">
        <div className="LoginForm__container padding-v-bigger bg-skew">
          <div className="container container-tiny">
            {this.props.authenticationError && (
              <div className="alert alert-error margin-c" role="alert">
                There was an error connecting to the Kubernetes API. Please check that your token is
                valid.
              </div>
            )}
          </div>
          <div className="bg-skew__pattern bg-skew__pattern-dark type-color-reverse">
            <div className="container">
              <h2>
                <Lock /> Login
              </h2>
              {this.props.oauthLoginURI ? this.oauthLogin() : this.tokenLogin()}
            </div>
          </div>
        </div>
      </section>
    );
  }

  private handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const { token } = this.state;
    return token && (await this.props.authenticate(this.props.cluster, token));
  };

  private handleTokenChange = (e: React.FormEvent<HTMLInputElement>) => {
    this.setState({ token: e.currentTarget.value });
  };

  private oauthLogin = () => {
    return (
      <>
        <p>
          Your cluster operator has enabled login via an authentication provider.{" "}
          <a
            href="https://github.com/kubeapps/kubeapps/blob/master/docs/user/using-an-OIDC-provider.md"
            target="_blank"
            rel="noopener noreferrer"
          >
            Click here
          </a>{" "}
          for more info about using authentication providers with Kubeapps.
        </p>
        <a href={this.props.oauthLoginURI} className="button button-accent">
          Login
        </a>
      </>
    );
  };

  private tokenLogin = () => {
    return (
      <>
        <p>
          Your cluster operator should provide you with a Kubernetes API token.{" "}
          <a
            href="https://github.com/kubeapps/kubeapps/blob/master/docs/user/access-control.md"
            target="_blank"
            rel="noopener noreferrer"
          >
            Click here
          </a>{" "}
          for more info on how to create and use Bearer Tokens.
        </p>
        <div className="bg-skew__content">
          <form onSubmit={this.handleSubmit}>
            <div>
              <label htmlFor="token">Kubernetes API Token</label>
              <input
                name="token"
                id="token"
                type="password"
                placeholder="Token"
                required={true}
                onChange={this.handleTokenChange}
                value={this.state.token}
              />
            </div>
            <p>
              <button type="submit" className="button button-accent">
                Login
              </button>
            </p>
          </form>
        </div>
      </>
    );
  };
}

export default LoginForm;
