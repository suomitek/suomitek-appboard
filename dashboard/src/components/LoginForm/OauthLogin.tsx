import * as React from "react";
import { CdsButton } from "../Clarity/clarity";

interface ILoginFormProps {
  authenticationError: string | undefined;
  oauthLoginURI: string;
}

function OAuthLogin(props: ILoginFormProps) {
  return (
    <section className="title" aria-labelledby="login-title" aria-describedby="login-desc">
      <h3 id="login-title" className="welcome">
        Welcome to <span>Suomitek-appboard</span>
      </h3>
      <p id="login-desc" className="hint">
        Your cluster operator has enabled login via an authentication provider.
      </p>
      <div className="login-group">
        {props.authenticationError && (
          <div className="error active">
            There was an error connecting to the Kubernetes API. Please check that your token is
            valid.
          </div>
        )}
        <a href={props.oauthLoginURI} className="login-submit-button">
          <CdsButton id="login-submit-button" status="primary">
            Login via OIDC Provider
          </CdsButton>
        </a>
      </div>
    </section>
  );
}

export default OAuthLogin;
