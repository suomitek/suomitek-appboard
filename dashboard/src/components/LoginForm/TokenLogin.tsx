import * as React from "react";
import { CdsButton } from "../Clarity/clarity";

interface ILoginFormProps {
  token: string;
  authenticationError: string | undefined;
  handleTokenChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

function TokenLogin(props: ILoginFormProps) {
  return (
    <section className="title" aria-labelledby="login-title" aria-describedby="login-desc">
      <h3 id="login-title" className="welcome">
        Welcome to <span>Suomitek-appboard</span>
      </h3>
      <p id="login-desc" className="hint">
        Your cluster operator should provide you with a Kubernetes API token.
      </p>
      <div className="login-group">
        <div className="clr-form-control">
          <label htmlFor="token" className="clr-control-label">
            Token
          </label>
          <div className="clr-control-container">
            <div className="clr-input-wrapper">
              <input
                type="password"
                id="token"
                placeholder="Paste token here"
                className="clr-input"
                required={true}
                onChange={props.handleTokenChange}
                value={props.token}
              />
            </div>
          </div>
        </div>
        {props.authenticationError && (
          <div className="error active ">
            There was an error connecting to the Kubernetes API. Please check that your token is
            valid.
          </div>
        )}
        <div className="login-submit-button">
          <CdsButton id="login-submit-button" status="primary">
            Submit
          </CdsButton>
        </div>
      </div>
    </section>
  );
}

export default TokenLogin;
