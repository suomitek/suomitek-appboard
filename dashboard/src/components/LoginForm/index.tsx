import * as React from "react";

import { ILoginFormProps } from "./LoginForm.v2";

const LoginForm = React.lazy(() => import("./LoginForm"));
const LoginFormV2 = React.lazy(() => import("./LoginForm.v2"));

interface ILoginFormSelectorProps extends ILoginFormProps {
  UI: string;
}

const LoginFormSelector: React.FC<ILoginFormSelectorProps> = props => (
  <React.Suspense fallback={null}>
    {props.UI === "clarity" ? <LoginFormV2 {...props} /> : <LoginForm {...props} />}
  </React.Suspense>
);

export default LoginFormSelector;
