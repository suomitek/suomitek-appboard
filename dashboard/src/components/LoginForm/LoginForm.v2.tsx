import { Location } from "history";
import * as React from "react";
import { useEffect, useState } from "react";
import { Redirect } from "react-router";
import { CdsIcon } from "../Clarity/clarity";

import LoadingWrapper from "../../components/LoadingWrapper";
import "./LoginForm.v2.css";
import OAuthLogin from "./OauthLogin";
import TokenLogin from "./TokenLogin";

export interface ILoginFormProps {
  cluster: string;
  authenticated: boolean;
  authenticating: boolean;
  authenticationError: string | undefined;
  oauthLoginURI: string;
  authenticate: (cluster: string, token: string) => any;
  checkCookieAuthentication: (cluster: string) => void;
  appVersion: string;
  location: Location;
}

function LoginForm(props: ILoginFormProps) {
  const [token, setToken] = useState("");
  const { oauthLoginURI, checkCookieAuthentication } = props;
  useEffect(() => {
    if (oauthLoginURI) {
      checkCookieAuthentication(props.cluster);
    }
  }, [oauthLoginURI, checkCookieAuthentication, props.cluster]);

  if (props.authenticating) {
    return <LoadingWrapper />;
  }
  if (props.authenticated) {
    const { from } = (props.location.state as any) || { from: { pathname: "/" } };
    return <Redirect to={from} />;
  }

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    return token && (await props.authenticate(props.cluster, token));
  };

  const handleTokenChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setToken(e.target.value);
  };

  return (
    <div className="login-wrapper">
      <form className="login clr-form" onSubmit={handleSubmit}>
        {props.oauthLoginURI ? (
          <OAuthLogin
            authenticationError={props.authenticationError}
            oauthLoginURI={props.oauthLoginURI}
          />
        ) : (
          <TokenLogin
            authenticationError={props.authenticationError}
            token={token}
            handleTokenChange={handleTokenChange}
          />
        )}
        <div className="login-moreinfo">
          <a
            href={`https://github.com/kubeapps/kubeapps/blob/${props.appVersion}/docs/user/access-control.md`}
            target="_blank"
            rel="noopener noreferrer"
          >
            <CdsIcon shape="info-circle" />
            More Info
          </a>
        </div>
      </form>
    </div>
  );
}

export default LoginForm;
