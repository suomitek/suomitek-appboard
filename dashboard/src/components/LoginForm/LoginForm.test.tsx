import { shallow } from "enzyme";
import { Location } from "history";
import context from "jest-plugin-context";
import * as React from "react";
import { Redirect } from "react-router";
import itBehavesLike from "../../shared/specs";

import LoginForm from "./LoginForm";

const emptyLocation: Location = {
  hash: "",
  pathname: "",
  search: "",
  state: "",
};

const defaultCluster = "default";

const defaultProps = {
  cluster: defaultCluster,
  authenticate: jest.fn(),
  authenticated: false,
  authenticating: false,
  authenticationError: undefined,
  location: emptyLocation,
  checkCookieAuthentication: jest.fn(),
  oauthLoginURI: "",
};

const authenticationError = "it's a trap";

describe("componentDidMount", () => {
  it("calls checkCookieAuthentication when oauthLoginURI provided", () => {
    const props = {
      ...defaultProps,
      oauthLoginURI: "/sign/in",
    };
    const checkCookieAuthentication = jest.fn();
    shallow(<LoginForm {...props} checkCookieAuthentication={checkCookieAuthentication} />);
    expect(checkCookieAuthentication).toHaveBeenCalled();
  });

  it("does not call checkCookieAuthentication when oauthLoginURI not provided", () => {
    const checkCookieAuthentication = jest.fn();
    shallow(<LoginForm {...defaultProps} checkCookieAuthentication={checkCookieAuthentication} />);
    expect(checkCookieAuthentication).not.toHaveBeenCalled();
  });
});

context("while authenticating", () => {
  itBehavesLike("aLoadingComponent", {
    component: LoginForm,
    props: { ...defaultProps, authenticating: true },
  });
});

describe("token login form", () => {
  it("renders a token login form", () => {
    const wrapper = shallow(<LoginForm {...defaultProps} />);
    expect(wrapper.find("input#token").exists()).toBe(true);
    expect(wrapper.find(Redirect).exists()).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders a link to the access control documentation", () => {
    const wrapper = shallow(<LoginForm {...defaultProps} />);
    expect(wrapper.find("a").props()).toMatchObject({
      href: "https://github.com/kubeapps/kubeapps/blob/master/docs/user/access-control.md",
      target: "_blank",
    });
  });

  it("updates the token in the state when the input is changed", () => {
    const wrapper = shallow(<LoginForm {...defaultProps} />);
    wrapper.find("input#token").simulate("change", { currentTarget: { value: "f00b4r" } });
    expect(wrapper.state("token")).toBe("f00b4r");
  });

  describe("redirect if authenticated", () => {
    it("redirects to / if no current location", () => {
      const wrapper = shallow(<LoginForm {...defaultProps} authenticated={true} />);
      const redirect = wrapper.find(Redirect);
      expect(redirect.props()).toEqual({ to: { pathname: "/" } });
    });

    it("redirects to previous location", () => {
      const location = Object.assign({}, emptyLocation);
      location.state = { from: "/test" };
      const wrapper = shallow(
        <LoginForm {...defaultProps} authenticated={true} location={location} />,
      );
      const redirect = wrapper.find(Redirect);
      expect(redirect.props()).toEqual({ to: "/test" });
    });
  });

  it("calls the authenticate handler when the form is submitted", () => {
    const wrapper = shallow(<LoginForm {...defaultProps} />);
    wrapper.find("input#token").simulate("change", { currentTarget: { value: "f00b4r" } });
    wrapper.find("form").simulate("submit", { preventDefault: jest.fn() });
    expect(defaultProps.authenticate).toBeCalledWith(defaultCluster, "f00b4r");
  });

  it("displays an error if the authentication error is passed", () => {
    const wrapper = shallow(
      <LoginForm {...defaultProps} authenticationError={authenticationError} />,
    );

    expect(wrapper.find(".alert-error").exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it("does not display the oauth login if oauthLoginURI provided", () => {
    const props = {
      ...defaultProps,
      oauthLoginURI: "",
    };

    const wrapper = shallow(<LoginForm {...props} />);

    expect(wrapper.find("a.button").exists()).toBe(false);
  });
});

describe("oauth login form", () => {
  const props = {
    ...defaultProps,
    oauthLoginURI: "/sign/in",
  };
  it("does not display the token login if oauthLoginURI provided", () => {
    const wrapper = shallow(<LoginForm {...props} />);

    expect(wrapper.find("input#token").exists()).toBe(false);
  });

  it("displays the oauth login if oauthLoginURI provided", () => {
    const wrapper = shallow(<LoginForm {...props} />);

    expect(wrapper.find("a.button").exists()).toBe(true);
  });

  it("renders a login button link", () => {
    const wrapper = shallow(<LoginForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
