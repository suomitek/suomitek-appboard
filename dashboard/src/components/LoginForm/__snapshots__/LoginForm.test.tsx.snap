// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oauth login form renders a login button link 1`] = `
<section
  className="LoginForm"
>
  <div
    className="LoginForm__container padding-v-bigger bg-skew"
  >
    <div
      className="container container-tiny"
    />
    <div
      className="bg-skew__pattern bg-skew__pattern-dark type-color-reverse"
    >
      <div
        className="container"
      >
        <h2>
          <Lock
            color="currentColor"
            size="24"
          />
           Login
        </h2>
        <p>
          Your cluster operator has enabled login via an authentication provider.
           
          <a
            href="https://github.com/kubeapps/kubeapps/blob/master/docs/user/using-an-OIDC-provider.md"
            rel="noopener noreferrer"
            target="_blank"
          >
            Click here
          </a>
           
          for more info about using authentication providers with Kubeapps.
        </p>
        <a
          className="button button-accent"
          href="/sign/in"
        >
          Login
        </a>
      </div>
    </div>
  </div>
</section>
`;

exports[`token login form displays an error if the authentication error is passed 1`] = `
<section
  className="LoginForm"
>
  <div
    className="LoginForm__container padding-v-bigger bg-skew"
  >
    <div
      className="container container-tiny"
    >
      <div
        className="alert alert-error margin-c"
        role="alert"
      >
        There was an error connecting to the Kubernetes API. Please check that your token is valid.
      </div>
    </div>
    <div
      className="bg-skew__pattern bg-skew__pattern-dark type-color-reverse"
    >
      <div
        className="container"
      >
        <h2>
          <Lock
            color="currentColor"
            size="24"
          />
           Login
        </h2>
        <p>
          Your cluster operator should provide you with a Kubernetes API token.
           
          <a
            href="https://github.com/kubeapps/kubeapps/blob/master/docs/user/access-control.md"
            rel="noopener noreferrer"
            target="_blank"
          >
            Click here
          </a>
           
          for more info on how to create and use Bearer Tokens.
        </p>
        <div
          className="bg-skew__content"
        >
          <form
            onSubmit={[Function]}
          >
            <div>
              <label
                htmlFor="token"
              >
                Kubernetes API Token
              </label>
              <input
                id="token"
                name="token"
                onChange={[Function]}
                placeholder="Token"
                required={true}
                type="password"
                value=""
              />
            </div>
            <p>
              <button
                className="button button-accent"
                type="submit"
              >
                Login
              </button>
            </p>
          </form>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`token login form renders a token login form 1`] = `
<section
  className="LoginForm"
>
  <div
    className="LoginForm__container padding-v-bigger bg-skew"
  >
    <div
      className="container container-tiny"
    />
    <div
      className="bg-skew__pattern bg-skew__pattern-dark type-color-reverse"
    >
      <div
        className="container"
      >
        <h2>
          <Lock
            color="currentColor"
            size="24"
          />
           Login
        </h2>
        <p>
          Your cluster operator should provide you with a Kubernetes API token.
           
          <a
            href="https://github.com/kubeapps/kubeapps/blob/master/docs/user/access-control.md"
            rel="noopener noreferrer"
            target="_blank"
          >
            Click here
          </a>
           
          for more info on how to create and use Bearer Tokens.
        </p>
        <div
          className="bg-skew__content"
        >
          <form
            onSubmit={[Function]}
          >
            <div>
              <label
                htmlFor="token"
              >
                Kubernetes API Token
              </label>
              <input
                id="token"
                name="token"
                onChange={[Function]}
                placeholder="Token"
                required={true}
                type="password"
                value=""
              />
            </div>
            <p>
              <button
                className="button button-accent"
                type="submit"
              >
                Login
              </button>
            </p>
          </form>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`while authenticating loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
  type={0}
/>
`;
